examples/hdf5.py

```diff
@@ -4,99 +4,184 @@
+the item is initialized so that the data is ready (or closer to ready) when
+the user needs it. This could alternatively be precomputed and stored in the HDF5 file.
+Cache sizes are limited, and a relatively small amount of cached data
+at the largest scales is required to fluidly navigate through large data sets.
 """
 
-import initExample ## Add path to library (just for examples; you do not need this)
+import initExample  # Add path to library (just for examples; you do not need this)
+
+import os
+import sys
+import math
+import threading
+from functools import lru_cache
 
-import sys, os
-import numpy as np
 import h5py
+import numpy as np
 import pyqtgraph as pg
 from pyqtgraph.Qt import QtCore, QtGui
 
 pg.mkQApp()
 
-
 plt = pg.plot()
 plt.setWindowTitle('pyqtgraph example: HDF5 big data')
 plt.enableAutoRange(False, False)
 plt.setXRange(0, 500)
 
-class HDF5Plot(pg.PlotCurveItem):
+
+class BigDataPlot(pg.PlotCurveItem):
     def __init__(self, *args, **kwds):
-        self.hdf5 = None
-        self.limit = 10000 # maximum number of samples to be plotted
+        self.sampler = None
         pg.PlotCurveItem.__init__(self, *args, **kwds)
         
-    def setHDF5(self, data):
-        self.hdf5 = data
-        self.updateHDF5Plot()
-        
+    def setDataSampler(self, sampler):
+        self.sampler = sampler
+        self.loadData()
+
     def viewRangeChanged(self):
-        self.updateHDF5Plot()
-        
-    def updateHDF5Plot(self):
-        if self.hdf5 is None:
+        self.loadData(lazy=True)
+
+    def loadData(self, lazy=False):
+        if self.sampler is None:
             self.setData([])
             return
-        
+
         vb = self.getViewBox()
         if vb is None:
             return  # no ViewBox yet
-        
-        # Determine what data range must be read from HDF5
-        xrange = vb.viewRange()[0]
-        start = max(0,int(xrange[0])-1)
-        stop = min(len(self.hdf5), int(xrange[1]+2))
-        
-        # Decide by how much we should downsample 
-        ds = int((stop-start) / self.limit) + 1
-        
-        if ds == 1:
-            # Small enough to display with no intervention.
-            visible = self.hdf5[start:stop]
-            scale = 1
+
+        x1, x2 = vb.viewRange()[0]
+        if not lazy or not self.sampler.isRangeValid(x1, x2):
+            x, y = self.sampler[x1:x2]
+            self.setData(x, y)  # update the plot
+
+
+class DataSampler:
+    def __init__(self, x, y, sampleLimit=2500, cacheLimit=10000000, padding=0.3, eager=True):
+        if x is not None and len(x) != len(y):
+            raise ValueError
+
+        self.x = x
+        self.y = y
+        self.sampleLimit = sampleLimit  # at most 4 times this number of samples will be plotted
+        self.cacheLimit = cacheLimit  # at most 8 times this number of samples will be cached
+        self.padding = padding
+        self._minCacheLevel = max(1, int(math.log2(max(1, int(len(self) / self.cacheLimit)))))
```

Comment:
when looking at `math.log2` of an int, I think you normally just want the `int.bit_length()`